### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] False positive for use of constant variable in loop initializer

### ISSUE THREAD

**Issue by TomerAberbach** [typescript-eslint/typescript-eslint#4413]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
const START = 2 as const

for (let i = START; i < 10; i++) {
  console.log(i)
}
```

```json
{
  "include": ["**/*.ts", "**/*.tsx"],
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ES2019"],
    "esModuleInterop": true,
    "jsx": "react-jsx",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "target": "ES2019",
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "~/*": ["app/*"]
    },
    "skipLibCheck": true,
    "noEmit": true
  }
}

```

**Expected Result**

No linting errors from `@typescript-eslint/no-unnecessary-condition`.

**Actual Result**

The following error is shown:
```
3:21  error  Unnecessary conditional, both sides of the expression are literal values  @typescript-eslint/no-unnecessary-condition
```

`3:21` refers to the `i < 10` expression.

At first I thought maybe the variable `i` was being typed as `let i: 2` (which would explain the linting error, because 2 is always less than 10), but TypeScript isn't showing any type errors for `i++` so I can only assume it's typing it as `let i: number`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.9.0` |
| `@typescript-eslint/parser`        | `5.9.0` |
| `TypeScript`                       | `4.5.4` |
| `ESLint`                           | `8.6.0` |
| `node`                             | `17.3.0` |

**Comment by bradzacher** [comment]
Inspecting the types - TS reports the type as `2`.

You can see the behaviour in the TS playground:
https://www.typescriptlang.org/play?#code/MYewdgzgLgBAygFQIICUEwLwwEwwIYQyiRQBQpAZiAE4wAUANgKawCWm8yaA3DOwDwwAjAAZerANQSAlDADepGEpgB6Fco2alAPQD8iouAghmAOgYgA5nVbTSAXyA

And you can inspect the types directly in this playground:
https://ts-ast-viewer.com/#code/MYewdgzgLgBAygFQIICUEwLwwEwwIYQyiRQBQpAZiAE4wAUANgKawCWm8yaA3DOwDwwAjAAZerANQSAlDADepGEXAQQzAHQMQAczqtppAL5A

We can only operate on the types that TS gives us, so this is "working as intended".

**Comment by TomerAberbach** [comment]
Huh interesting. Would you consider this a bug in TypeScript? It seems wrong for TS to make the type `2` but allow you to write `i++`

**Comment by bradzacher** [comment]
Not really, no.
For `let` assignments TS starts with as narrow of a type as possible and broadens the type as applicable as you assign and reassign it.

As far as TS is concerned - that location of the control flow analysis of the code the variable has not been reassigned yet.

One could argue that TS should analyse the context (that it's a loop variable) and that the type should be widened by default - but it's likely not something that's worth spending the time to fix.
